Fleshing out README

ycombinator opened the pull request:
This PR fleshes out the README by adding a list of supported commands under `elastic-package` and short descriptions for each.

Preview of rendered README: https://github.com/ycombinator/elastic-package/blob/readme/README.md.

Resolves partially #75.

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-76/detail/PR-76/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-76/detail/PR-76/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-76/detail/PR-76/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-76/detail/PR-76/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-package_76.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #76 opened]`

* Start Time: 2020-08-27T04:41:11.462+0000

* Duration: 4 min 48 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->





<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ycombinator commented on README.md:
```diff
@@ -24,18 +17,80 @@ Change directory to the package under development. Note: an integration is a spe
+### `elastic-package build`
+
+Use this command to build a package. TODO: what is the purpose of building a package — when would
+a package developer want/need to run this command?
```
@mtojek Could you help answer these TODO questions here, please? Thanks!

mtojek replied:
The `build` command copies package internals to the build directory and encode Kibana dashboards. Now it will also skip `_dev` directory.

ycombinator replied:
Thanks, I understand what it does internally. I was more wondering why a package developer would want or need to run this command? What use case does it support? Maybe it's supposed to be a pre-requisite for some other commands or something like that?

mtojek replied:
I'm ok with both options, let's wait with announcing them and silently observe in the wild first.

ycombinator replied:
Sorry, but I didn't understand. Did you mean to post your answer to a different comment thread?

mtojek replied:
> Sorry, but I didn't understand. Did you mean to post your answer to a different comment thread?

Yes, I had poor network connection. I (or JS) must have messed up sth with anchors. 

Anyway, this is good question. I agree with you that there is no direct goal of running `build` command. Currently built artifacts are used to:
1. Build docker image consisting of package-registry with custom packages.
2. Push changes from integrations to the package-storage.
3. Dry-run package-registry (I believe we should move towards package-spec, as that was an ultimate goal). @ruflin do you know if there are any features that block switching to spec validation?

ruflin replied:
Not now, but in the future I expect this to build the `.zip` file of the package. 

++ on switching to package spec validation and remove the validation code itself from the registry. Each validation part we remove we can check if it is also in the spec.

ruflin replied:
Additional note: We should still have the validation in the package-registry but run it based on package-spec. So far some time we will probably do double validation but this should not be an issue.

ycombinator replied:
Thanks for the clarifications @mtojek. 

> 1. Build docker image consisting of package-registry with custom packages.
> 2. Push changes from integrations to the package-storage.

I have added some text covering these two use cases.

> 3. Dry-run package-registry.

I don't understand what this means, how it's different from 1., and how locally-built packages help here, particularly from a package author's perspective. It sounds like this is an internal use, not something package authors need to care about. So I have not mentioned this in the README text for this command.

ycombinator commented on README.md:
```diff
@@ -24,18 +17,80 @@ Change directory to the package under development. Note: an integration is a spe
+
+### `elastic-package promote`
+
+Use this command to promote a package to the Snapshot Package Registry.
```
@mtojek Is this description accurate?

mtojek replied:
The promotion from Integrations to package-storage happens automagically (Jenkins job).

The `promote` command exposes a package selection form in the TUI, which simplifies promotion between different stages (snapshot, staging, production) including opening relevant pull requests to the package-storage.

ycombinator replied:
Ah yes, thanks for reminding me! I can fix the description. Is this command one we want to document or is it meant more for administrators only? We can also document it but say that it's not meant for general use or something like that.

mtojek replied:
Hm, I wouldn't keep it secret (do not mention in README), but it's valid point to add a notice about administrator use only.

ycombinator replied:
Sounds good, thanks. I will add a notice about administrator use only.

ycombinator replied:
Added in 291401e.

ycombinator commented on README.md:
```diff
@@ -24,25 +17,89 @@ Change directory to the package under development. Note: an integration is a spe
-Make sure you have access to public repositories (to open pull requests) and to user data (check authorized account data).
+Make sure you have enabled the following scopes:
+* `public_repo` — to open pull requests on GitHub repositories.
+* `read:user` and `user:email` — to read your user profile information from GitHub in order to populate pull requests appropriately.
```
@mtojek Could you confirm that this is accurate, please? Thanks!

mtojek replied:
I'm not convince if user:email is required, but let's keep it if we need this in the future.

mtojek commented on README.md:
```diff
@@ -24,25 +17,89 @@ Change directory to the package under development. Note: an integration is a spe
+
+### `elastic-package lint`
+
+Use this command to validate the contents of a package.
```
... against schema files defined by package-spec repository.

ycombinator replied:
Referenced the package spec in 51c497473dffecb4d6163351d23d6b2e5bcb5e25.

mtojek commented on README.md:
```diff
@@ -24,25 +17,89 @@ Change directory to the package under development. Note: an integration is a spe
+
+### `elastic-package stack`
+
+Use this command to spin up a Docker-based Elastic Stack consisting of Elasticsearch, Kibana, and 
```
typo: s/, and/ and/

ycombinator replied:
No, this was intentional per:

![image](https://user-images.githubusercontent.com/51061/91462717-58e1d400-e83f-11ea-90a2-44ce5749d5d1.png)

mtojek replied:
No more comments, thank you for the explanation :D

mtojek commented on README.md:
```diff
@@ -24,25 +17,89 @@ Change directory to the package under development. Note: an integration is a spe
+### `elastic-package stack`
+
+Use this command to spin up a Docker-based Elastic Stack consisting of Elasticsearch, Kibana, and 
+the Package Registry.
```
It supports different versions of the stack?

ycombinator replied:
Yes, generally I was trying to avoid mentioning all the flags for each command but I agree, this one is good to mention. Will add!

ycombinator replied:
Added in 0d6b1e6.

mtojek commented on README.md:
```diff
@@ -1,16 +1,9 @@
+Currently, `elastic-package` only supports packages of type [Elastic Integrations](https://github.com/elastic/integrations).
 
 ## Getting started
 
```
If you are cleaning up the README, it might be a good option to refresh the Getting started as well. Currently, with `go get` we will loose the information about the version, so maybe it's better to replace it with `git clone` and `make build`.

We can also mention that the tool is go-gettable (but it will loose information about revision).

ycombinator replied:
Good point — I will clarify this in the README.

However, in the future (not for this PR), I hope we can have a single command to install `elastic-package` with all it's features. It seems weird from a user's perspective that the `version` command only works if you checkout the source and build from it (but I understand why that is the case currently). In the future, maybe we can think of committing a `version.go` file that's updated every so often via `make build` and then `elastic-package version` reads from this file or something like that.

ycombinator replied:
Updated Getting Started section in efe6362dd89696e60eb48eb4dd845a67581eea76.

mtojek commented on README.md:
````diff
@@ -24,25 +17,89 @@ Change directory to the package under development. Note: an integration is a spe
 ```
 
-## GitHub authorization
+## Commands
````
Do you think we can sort these commands and maybe keep help at the top of them?

ycombinator replied:
Sorted in 51c497473dffecb4d6163351d23d6b2e5bcb5e25.

mtojek commented on README.md:
````diff
@@ -24,25 +17,89 @@ Change directory to the package under development. Note: an integration is a spe
 ```
 
-## GitHub authorization
+## Commands
````
Did you miss the test command or is it intended to cover it later on once system tests are done?

ycombinator replied:
I intended to add it when system tests are done but, now that you mention it, we have pipeline tests so maybe we should add it to the README now itself? Or would you prefer to wait until we have defined pipeline test files in a couple of integrations packages first?

mtojek replied:
I'm fine with both options, we can wait with announcing them officially.

ycombinator replied:
I'm going to wait to add the `test` command here until https://github.com/elastic/integrations/pull/262 is merged. We can easily update the README at that time.

mtojek replied:
FYI PR is merged.

ycombinator replied:
Great! I will add `test` here now.

ycombinator replied:
Added a section in bd93e6a. 

But I think we need to now define what a Pipeline Test is and how a package author can set it up. Let's do this in a follow up PR. I added a checklist item for this in https://github.com/elastic/elastic-package/issues/75. @mtojek Since you worked on the pipeline tests, can you take on this follow up PR?

mtojek replied:
HOWTO guides is a good idea. I can write some docs about it.

mtojek reviewed (commented):
Ah, finally a decent README :) Thanks for writing it. I left a couple of comments on various aspects.

ycombinator commented:
@mtojek Thanks for the review. I believe I've addressed all your feedback but some of the comments need more discussion. Please re-review when you get a chance. Thanks!

mtojek commented on README.md:
```diff
@@ -64,30 +58,28 @@ a package developer want/need to run this command?
-Use this command to validate the contents of a package.
+### `elastic-package check`
+
+Use this command to run the `format`, `lint`, and `build` commands all at once, in that order.
```
... and the future the "test" subcommand.

I wonder if we should include "test" in the "check" chain as it requires the cluster to be up and running.

ycombinator replied:
Let's keep the README about current state as much as possible. We can always update it easily as/when we make enhancements or changes.

I could go either way on including `test` in `check` but I slightly favor keeping it out. This way `check` can run relatively fast and can be run as an earlier stage in CI.

mtojek reviewed (approved):
Only few nit-picks. LGTM

ycombinator commented on README.md:
```diff
@@ -24,25 +24,99 @@ Change directory to the package under development. Note: an integration is a spe
+
+Use this command to promote a package to the Snapshot Package Registry.
+
+_Context: package_
```
@mtojek I think this should be `global`, right, since the tool lets the user select multiple packages at once?

mtojek replied:
Correct, I missed this one. It should be `global`.

ycombinator replied:
Fixed in 291401e.

ycombinator commented:
Hey @mtojek thanks for the re-review. I've addressed the latest feedback; please take a look. The main item that I need your input on is https://github.com/elastic/elastic-package/pull/76#discussion_r478504388.

mtojek reviewed (approved):
Re-reviewed, nothing blocking here!

ycombinator commented:
I think, at this point, the contents of this README PR are already going to be helpful to package authors so I'm going to merge it.

There are a couple of rough edges that we might want to polish, but I think those can be done in follow up PRs (see https://github.com/elastic/elastic-package/pull/76#discussion_r480276789 and https://github.com/elastic/elastic-package/pull/76#discussion_r480273409).
